### ISSUE TITLE
[BUG] std on dask_cudf frame fails using the agg api 

### ISSUE THREAD

**Issue by VibhuJawa** [rapidsai/cudf#4388]
**Describe the bug**
`std` on `dask_cudf` frame fails using the `agg` api 

**Steps/Code to reproduce bug**
```python
import cudf
import dask_cudf

df = cudf.DataFrame({'a': [1,1,2,2],'b': [4,5,6,10]})
ddf = dask_cudf.from_cudf(df, npartitions=2)

ddf.groupby('a').agg({'b':['mean','std']}).compute()
```
#### Trace:
```python
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    171     try:
--> 172         yield
    173     except Exception as e:

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   4937     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 4938         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   4939 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py in _groupby_apply_funcs(df, *index, **kwargs)
    880     for result_column, func, func_kwargs in funcs:
--> 881         r = func(grouped, **func_kwargs)
    882 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py in _compute_sum_of_squares(grouped, column)
    898     base = grouped[column] if column is not None else grouped
--> 899     return base.apply(lambda x: (x ** 2).sum())
    900 

AttributeError: 'SeriesGroupBy' object has no attribute 'apply'

During handling of the above exception, another exception occurred:

ValueError                                Traceback (most recent call last)
<ipython-input-13-ae54f4074622> in <module>
      5 ddf = dask_cudf.from_cudf(df, npartitions=2)
      6 
----> 7 ddf.groupby('a').agg({'b':['mean','std']}).compute()

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py in agg(self, arg, split_every, split_out)
   1756     @derived_from(pd.core.groupby.DataFrameGroupBy)
   1757     def agg(self, arg, split_every=None, split_out=1):
-> 1758         return self.aggregate(arg, split_every=split_every, split_out=split_out)
   1759 
   1760 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1751 
   1752         return super(DataFrameGroupBy, self).aggregate(
-> 1753             arg, split_every=split_every, split_out=split_out
   1754         )
   1755 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py in aggregate(self, arg, split_every, split_out)
   1532             split_out=split_out,
   1533             split_out_setup=split_out_on_index,
-> 1534             sort=self.sort,
   1535         )
   1536 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/core.py in apply_concat_apply(args, chunk, aggregate, combine, meta, token, chunk_kwargs, aggregate_kwargs, combine_kwargs, split_every, split_out, split_out_setup, split_out_setup_kwargs, sort, **kwargs)
   4891 
   4892     if meta is no_default:
-> 4893         meta_chunk = _emulate(chunk, *args, udf=True, **chunk_kwargs)
   4894         meta = _emulate(aggregate, _concat([meta_chunk]), udf=True, **aggregate_kwargs)
   4895     meta = make_meta(

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/core.py in _emulate(func, *args, **kwargs)
   4936     """
   4937     with raise_on_meta_error(funcname(func), udf=kwargs.pop("udf", False)):
-> 4938         return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
   4939 
   4940 

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/contextlib.py in __exit__(self, type, value, traceback)
    128                 value = type()
    129             try:
--> 130                 self.gen.throw(type, value, traceback)
    131             except StopIteration as exc:
    132                 # Suppress StopIteration *unless* it's the same exception that

/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/utils.py in raise_on_meta_error(funcname, udf)
    191         )
    192         msg = msg.format(" in `{0}`".format(funcname) if funcname else "", repr(e), tb)
--> 193         raise ValueError(msg)
    194 
    195 

ValueError: Metadata inference failed in `_groupby_apply_funcs`.

You have supplied a custom function and Dask is unable to 
determine the type of output that that function returns. 

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
AttributeError("'SeriesGroupBy' object has no attribute 'apply'")

Traceback:
---------
  File "/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/utils.py", line 172, in raise_on_meta_error
    yield
  File "/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/core.py", line 4938, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 881, in _groupby_apply_funcs
    r = func(grouped, **func_kwargs)
  File "/raid/vjawa/miniconda3/envs/cudf_march_9/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 899, in _compute_sum_of_squares
    return base.apply(lambda x: (x ** 2).sum())
```

**Expected behaviour**
```python
import cudf
import dask_cudf

df = cudf.DataFrame({'a': [1,1,2,2],'b': [4,5,6,10]})
ddf = dask_cudf.from_cudf(df, npartitions=2).to_dask_dataframe()

ddf.groupby('a').agg({'b':['mean','std']}).compute()
```
```
b
mean	std
a		
1	4.5	0.707107
2	8.0	2.828427
```

**Environment details**
```
# packages in environment at /raid/vjawa/miniconda3/envs/cudf_march_9:
cudf                      0.13.0a200309         py37_3294    rapidsai-nightly
dask-cudf                 0.13.0a200309         py37_3294    rapidsai-nightly
libcudf                   0.13.0a200309     cuda10.1_3294    rapidsai-nightly
dask                      2.12.0                     py_0    conda-forge
dask-core                 2.12.0                     py_0    conda-forge
dask-cuda                 0.13.0b200309           py37_65    rapidsai-nightly
dask-cudf                 0.13.0a200309         py37_3294    rapidsai-nightly
```

**Comment by beckernick** [comment]
Hopefully, a duplicate of https://github.com/rapidsai/cudf/issues/3429

**Comment by VibhuJawa** [comment]
> Hopefully, a duplicate of #3429

I dont think it is though, If i am reading the code correctly, we never call `.std` on dask-dataframe rather it does these 3 things:

It calculates below for each group:  
- `sum`
- `count`
-  `sum of squares`

For formula see : http://mathforum.org/library/drmath/view/61381.html

For dask code see: 

https://github.com/dask/dask/blob/c8cda684fda06480d32e8e62c78080e089b4f04f/dask/dataframe/groupby.py#L938-L949

So for this to work we need to have a `sum of squares` function for each group , i.e port the below function to work with cudf .  

https://github.com/dask/dask/blob/c8cda684fda06480d32e8e62c78080e089b4f04f/dask/dataframe/groupby.py#L897-L900

**Comment by beckernick** [comment]
Thanks for digging into this. From the above it sounds like the lack of API surface is blocking this, but we would then fail downstream due to what you've found.

cc @shwina @kkraus14 for visibility on the internal function used for the blocked standard deviation calculation

**Comment by beckernick** [comment]
Actually, I believe the `groupby.apply(lambda x: (x ** 2).sum())` may "just work" once the groupby refactor lands

**Comment by kkraus14** [comment]
> Actually, I believe the `groupby.apply(lambda x: (x ** 2).sum())` may "just work" once the groupby refactor lands

That will be MUCH slower than a native aggregation where I'd highly recommend against it.

cc @quasiben could you share the CustomAggregation path here?

**Comment by quasiben** [comment]
Might require some a tiny bit of cudf/dask-cudf work or maybe just some _meta munging

- https://docs.dask.org/en/latest/dataframe-groupby.html#aggregate
- https://gist.github.com/quasiben/e1a7083d0cf45825cecd76649726459a

**Comment by kkraus14** [comment]
Was this resolved by #4346 ?

**Comment by VibhuJawa** [comment]
```python
>>> import cudf
>>> import dask_cudf
>>> df = cudf.DataFrame({'a': [1,1,2,2],'b': [4,5,6,10]})
>>> ddf = dask_cudf.from_cudf(df, npartitions=2)
>>> ddf.groupby('a').agg({'b':['mean','std']}).compute()
```
Now we get below error : 

```python
Traceback (most recent call last):
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/utils.py", line 172, in raise_on_meta_error
    yield
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/core.py", line 4938, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 881, in _groupby_apply_funcs
    r = func(grouped, **func_kwargs)
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 899, in _compute_sum_of_squares
    return base.apply(lambda x: (x ** 2).sum())
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/cudf/core/groupby/groupby.py", line 336, in apply
    result = cudf.concat([function(chk) for chk in chunks])
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/cudf/core/reshape.py", line 106, in concat
    raise ValueError("Unknown type %r" % typ)
ValueError: Unknown type <class 'numpy.int64'>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1758, in agg
    return self.aggregate(arg, split_every=split_every, split_out=split_out)
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1753, in aggregate
    arg, split_every=split_every, split_out=split_out
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 1534, in aggregate
    sort=self.sort,
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/core.py", line 4893, in apply_concat_apply
    meta_chunk = _emulate(chunk, *args, udf=True, **chunk_kwargs)
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/core.py", line 4938, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/contextlib.py", line 130, in __exit__
    self.gen.throw(type, value, traceback)
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/utils.py", line 193, in raise_on_meta_error
    raise ValueError(msg)
ValueError: Metadata inference failed in `_groupby_apply_funcs`.

You have supplied a custom function and Dask is unable to 
determine the type of output that that function returns. 

To resolve this please provide a meta= keyword.
The docstring of the Dask function you ran should have more information.

Original error is below:
------------------------
ValueError("Unknown type <class 'numpy.int64'>")

Traceback:
---------
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/utils.py", line 172, in raise_on_meta_error
    yield
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/core.py", line 4938, in _emulate
    return func(*_extract_meta(args, True), **_extract_meta(kwargs, True))
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 881, in _groupby_apply_funcs
    r = func(grouped, **func_kwargs)
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/dask/dataframe/groupby.py", line 899, in _compute_sum_of_squares
    return base.apply(lambda x: (x ** 2).sum())
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/cudf/core/groupby/groupby.py", line 336, in apply
    result = cudf.concat([function(chk) for chk in chunks])
  File "/raid/vjawa/conda_install/conda_env/envs/cudf_march_18_3/lib/python3.7/site-packages/cudf/core/reshape.py", line 106, in concat
    raise ValueError("Unknown type %r" % typ)
```

**Comment by jakirkham** [comment]
cc @rjzamora (for awareness)

**Comment by shwina** [comment]
Seems that the problem is that `cudf.concat()` above is called on scalars rather than cudf objects:

```
In [10]: cudf.concat([np.int64(0), np.int64(1)])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-10-e64692f93d32> in <module>
----> 1 cudf.concat([np.int64(0), np.int64(1)])

~/workspace/github/shwina/cudf/python/cudf/cudf/core/reshape.py in concat(objs, axis, ignore_index, sort)
    104         return Index._concat(objs)
    105     else:
--> 106         raise ValueError("Unknown type %r" % typ)
    107
    108

ValueError: Unknown type <class 'numpy.int64'>
```

Should we just support concatenating scalars into a Series via `cudf.concat`? @kkraus14

**Comment by kkraus14** [comment]
Are we returning scalars somewhere that we shouldn't be?

**Comment by rgsl888prabhu** [comment]
> Are we returning scalars somewhere that we shouldn't be?

Scalar is returned by [sum](https://github.com/rapidsai/cudf/blob/85c75c2fc69ce075e1ca504850d3d3ba10e918c5/python/cudf/cudf/core/column/numerical.py#L180) which is called by [_compute_sum_of_squares](https://github.com/dask/dask/blob/a7c7446bbfb2b344e07e1a8dbec6db33b69a39a9/dask/dataframe/groupby.py#L897)

**Comment by rgsl888prabhu** [comment]
simple repro
```
import cudf

df = cudf.DataFrame({'a':[0, 1], 'b':[0, 1]})

print(df.groupby('a').b.apply(lambda x: (x ** 2).sum()))
```

**Comment by kkraus14** [comment]
> simple repro
> 
> ```
> import cudf
> 
> df = cudf.DataFrame({'a':[0, 1], 'b':[0, 1]})
> 
> print(df.groupby('a').b.apply(lambda x: (x ** 2).sum()))
> ```

While we should enable this to work, this is a much slower path that we should try to push changes upstream into Dask to avoid the `apply` iterative path.

**Comment by rjzamora** [comment]
Just a note that [dask#6186](https://github.com/dask/dask/pull/6186) should address this.

**Comment by rjzamora** [comment]
Closing this since [dask#6186](https://github.com/dask/dask/pull/6186) is merged.
